### PR TITLE
[tex] enhance locking and pause chef-client daemon when applying proposals

### PR DESCRIPTION
**Pull request by aspiers**
Backport of https://github.com/crowbar/crowbar-core/pull/14 to tex:

Reworking of https://github.com/crowbar/crowbar-core/pull/5 to address concurrency issues, make the interface between the two locking subclasses consistent, and DRY up the calling code.

It requires and is required by https://github.com/crowbar/barclamp-network/pull/330, so it will be presumably impossible for the gate to pass either of these PRs.

### CODE REVIEW

File: crowbar_framework/app/models/crowbar_service.rb [hunk 1]
```diff
@@ -47,9 +47,10 @@ def transition(inst, name, state)
 
     @logger.info("Crowbar transition enter: #{name} to #{state}")
 
-    f = acquire_lock "BA-LOCK"
-    begin
-      node = NodeObject.find_node_by_name name
+    pop_it = false
+    node = NodeObject.find_node_by_name name
```
Comment (vuntz) [inline crowbar_framework/app/models/crowbar_service.rb]: Missing the latest fix from the original PR.

Comment (aspiers) [inline crowbar_framework/app/models/crowbar_service.rb]: Thanks; fixed.

**Comment by vuntz** [comment]
+1

**Comment by aspiers** [comment]
Gate failure: `crowbar-testbuild.py` is causing `Build failed: <class 'sh.ErrorReturnCode_1'>`.  Have added to https://etherpad.nue.suse.com/p/cloud-gating (can't deeplink unless we have a #TrelloGatingWhiteboard )

**Comment by aspiers** [comment]
Issue magically vanished after a forced rebuild :-(  I suspect a lack of error reporting in `crowbar-testbuild.py`, which https://github.com/SUSE-Cloud/automation/pull/466 should address.

**Comment by aspiers** [comment]
Manually kicked off build with [manual merge](http://tu-sle12.j.cloud.suse.de/mkcloud/tex-new-locking/) of test packages for this PR and https://github.com/crowbar/barclamp-network/pull/330.

**Comment by aspiers** [comment]
[Last manual build failed](https://ci.suse.de/job/openstack-mkcloud/7951/) for some weird reason which looks unrelated so I've forced another rebuild which will be discoverable by searching for `barclamp-crowbar:1353` in https://ci.suse.de/job/openstack-mkcloud/buildTimeTrend

**Pull request merged**